include/revng-c/mlir/Dialect/Clift/IR/CliftStorage.h

```diff
@@ -12,166 +14,109 @@
-    if (TheKey.fields.has_value() and body == *TheKey.fields
-        and TheKey.name == name and TheKey.Size == Size)
-      return mlir::success();
-
-    if (TheKey.fields.has_value())
-      return mlir::failure();
-
-    TheKey.fields = llvm::SmallVector<FieldAttr, 2>();
-    for (auto field : body)
-      TheKey.fields->push_back(field);
-    TheKey.name = alloc.copyInto(name);
-    TheKey.Size = Size;
-    return mlir::success();
-  }
-
-  [[nodiscard]] llvm::StringRef getName() const { return TheKey.name; }
-
-  [[nodiscard]] bool isInitialized() const { return TheKey.isInitialized(); }
-
-  llvm::ArrayRef<FieldAttr> getFields() const {
-    revng_assert(isInitialized());
-    return *TheKey.fields;
-  }
-
-  uint64_t getSize() const { return TheKey.Size; }
+    [[nodiscard]] llvm::hash_code hashValue() const {
+      return llvm::hash_value(ID);
+    }
+  };
 
-  uint64_t getID() const { return TheKey.ID; }
+  // Some internal mechanism of mlir have access to the key of the storage
+  // instead of the entire storage or the key using that storage. For that
+  // reason it is best to throw everything related to the type inside the key,
+  // and then override operators to pretend they don't know about the non key
+  // fields. That way when it is needed one can access everything.
+  struct Key : KeyBase, ValueT {
+    template<typename... ArgsT>
+    Key(const uint64_t ID, const llvm::StringRef Name, ArgsT &&...Args) :
+      KeyBase{ ID, Name, std::nullopt },
+      ValueT{ std::forward<ArgsT>(Args)... } {}
+  };
 
-private:
   Key TheKey;
-};
 
-struct UnionTypeStorage : public mlir::AttributeStorage {
 public:
-  struct Key {
-
-    uint64_t ID;
-    llvm::StringRef name;
-    Optional<llvm::SmallVector<FieldAttr, 2>> fields;
-
-    bool operator==(const Key &Other) const { return Other.ID == ID; }
-
-    llvm::hash_code hashValue() const { return llvm::hash_value(ID); }
-
-    [[nodiscard]] bool isInitialized() const { return fields.has_value(); }
-  };
-
   using KeyTy = Key;
 
-  static llvm::hash_code hashKey(const KeyTy &key) { return key.hashValue(); }
+  static llvm::hash_code hashKey(const KeyTy &Key) { return Key.hashValue(); }
 
-  UnionTypeStorage(uint64_t ID) { TheKey.ID = ID; }
+  bool operator==(const Key &Other) const { return TheKey == Other; }
 
-  /// Define the comparison function.
-  bool operator==(const KeyTy &key) const { return key == TheKey; }
-
-  /// Define a construction method for creating a new instance of the storage.
-  static UnionTypeStorage *
-  construct(mlir::StorageUniquer::StorageAllocator &allocator,
-            const KeyTy &Key) {
-    auto ToReturn = new (allocator.allocate<UnionTypeStorage>())
-      UnionTypeStorage(Key.ID);
-    if (Key.isInitialized()) {
-      auto Res = ToReturn->mutate(allocator, Key.name, *Key.fields);
-      revng_assert(Res.succeeded());
-    }
+  template<typename... ArgsT>
+  ClassTypeStorage(const uint64_t ID,
+                   const llvm::StringRef Name,
```

Comment:
Just FYI, `llvm::StringRef` already has the semantic of an immutable reference, so `const` is not really strictly necessary.

Reply:
`const` here applies to the `StringRef` object, preventing its reassignment, and not to the `char`s to which it refers.
`const int* p;` vs `int* const p;`
This is really just a personal preference thing. I always mark as const any locals that are not mutated. Improves correctness and readability.

Reply:
Yeah sure. I have the same preference. I just took the opportunity to note this fact about `StringRef` being an immutable reference, which kinda surprised me the first time I realized it.